### ISSUE TITLE
Leaky_relu exploding values

### ISSUE THREAD

**Issue by lodykas** [neuronika/neuronika#100]
Hello, 

Basically, The following code produces exploding gradient, values and losses. I've tested with different optimizers (SGD, Adagrad, Adam) for multiple activation functions (softplus, relu and leaky_relu).

Optimisers parameters change the time to convergence, but basically, while softplus and relu stabilizes around the correct answer with relatively small losses, waiting for a sufficiently large number of epochs with lead to exploding values with leaky_relu:

```
[...]
epoch 181 loss: 18083050
epoch 182 loss: 14513438
epoch 183 loss: 211705540
epoch 184 loss: 113424490
epoch 185 loss: 4869084000
epoch 186 loss: 981610800
epoch 187 loss: 225812140000
epoch 188 loss: 2468554200000
epoch 189 loss: 131793980000000
epoch 190 loss: 24977954000000000
epoch 191 loss: 88780490000000000000
epoch 192 loss: 138728520000000000000000000
epoch 193 loss: 105674000000000000000000000000000000000
epoch 194 loss: NaN
epoch 195 loss: NaN
epoch 196 loss: NaN
epoch 197 loss: NaN
epoch 198 loss: NaN
epoch 199 loss: NaN
[...]
```

I'm not sure how to  completely seed the learning process, but I can make this behavior happen consistently. ReLU and Softplus should offer relatively similar performances I think, so there probably is some kind of computational mistake here ( I assume it's not my code since I've tested many things)

``` Rust
use ndarray::{Ix4, stack, Axis, Array, Dim};

use neuronika::Param;
use neuronika::{Data, Gradient, Overwrite, VarDiff, Convolve, from_ndarray};
use neuronika::optim::{Adam, L2};
use neuronika::nn::{ModelStatus, Conv2d, Learnable};
use neuronika::nn;
use neuronika::nn::init::{calculate_gain, xavier_normal};

// Network definition.
pub struct NeuralNetwork {
	con1: Conv2d<nn::Zero>,
	con2: Conv2d<nn::Zero>,
	con3: Conv2d<nn::Zero>,
	con4: Conv2d<nn::Zero>,
	status: ModelStatus,     
}

impl NeuralNetwork {
	fn new() -> Self {
		let mut status = ModelStatus::default();

		let con1 = status.register(
			Conv2d::new(5, 20, (5, 5), (0, 0), nn::Zero, (1, 1), (1, 1))
		);
		let con2 = status.register(
			Conv2d::new(20, 40, (5, 5), (0, 0), nn::Zero, (1, 1), (1, 1))
		);
		let con3 = status.register(
			Conv2d::new(40, 20, (2, 2), (0, 0), nn::Zero, (1, 1), (1, 1))
		);
		let con4 = status.register(
			Conv2d::new(20, 1, (1, 1), (0, 0), nn::Zero, (1, 1), (1, 1))
		);
		
		Self {
			con1: con1,
			con2: con2,
			con3: con3,
			con4: con4,
			status,
		}
	}

    fn forward<I, T, U>(
        &self,
        input: I,
    ) -> VarDiff<impl Data<Dim = Ix4>, impl Gradient<Dim = Ix4> + Overwrite>
    where
		I: Convolve<I, Learnable<Ix4>, nn::Zero>,
		I::Output: Into<VarDiff<T, U>>,
		T: Data<Dim = Ix4>,
		U: Gradient<Dim = Ix4> + Overwrite,
    {
		let out1 = self.con1.forward(input).leaky_relu();
		let out2 = self.con2.forward(out1).leaky_relu();
		let out3 = self.con3.forward(out2).leaky_relu();
		let out4 = self.con4.forward(out3);
		out4
    }


	pub fn parameters(&self) -> Vec<Param> {
		self.status.parameters()
	}

	/// Switches the network in training mode.
	fn train(&self) {
		self.status.train()
	}
}


fn main(){
	let model = NeuralNetwork::new();
	let model_optim = neuronika::optim::SGD::new(model.parameters(), 0.001, neuronika::optim::L2::new(0.));
	model.train();
	
	// input data should be irrevelant here (see target below)
	let data_point: Array<f32, Dim<[usize; 4]>> = stack(Axis(0), &[Array::eye(10).view(), Array::eye(10).view(), Array::eye(10).view(), Array::eye(10).view(), Array::eye(10).view()]).unwrap().into_shape((1, 5, 10, 10)).unwrap(); 
	
	for i in 0..600{
	    let x = model.forward(from_ndarray(data_point.clone()));
	    
	    // model should learn constant function f(x) = 10
	    let y = from_ndarray(Array::from_vec(vec![10.]).into_shape((1, 1, 1, 1)).unwrap());
	    
	    let loss = nn::loss::mae_loss(x.clone(), y, nn::loss::Reduction::Mean);
	    loss.forward();
	    loss.backward(1.);
	    
	    println!("epoch {} loss: {}", i, &loss.data());
	    
	    model_optim.step();
	}
	
}
```

**Comment by frjnn** [comment]
Hi, you probably forgot to zero gradients between each epoch. This is a bit our fault too since in the example we do not show it.

```rust
use ndarray::{stack, Array, Axis, Dim, Ix4};

use neuronika::nn;
use neuronika::nn::init::{calculate_gain, xavier_normal};
use neuronika::nn::{Conv2d, Learnable, ModelStatus};
use neuronika::optim::{Adam, L2};
use neuronika::Param;
use neuronika::{from_ndarray, Convolve, Data, Gradient, Overwrite, VarDiff};

// Network definition.
pub struct NeuralNetwork {
    con1: Conv2d<nn::Zero>,
    con2: Conv2d<nn::Zero>,
    con3: Conv2d<nn::Zero>,
    con4: Conv2d<nn::Zero>,
    status: ModelStatus,
}

impl NeuralNetwork {
    fn new() -> Self {
        let mut status = ModelStatus::default();

        let con1 = status.register(Conv2d::new(5, 20, (5, 5), (0, 0), nn::Zero, (1, 1), (1, 1)));
        let con2 = status.register(Conv2d::new(
            20,
            40,
            (5, 5),
            (0, 0),
            nn::Zero,
            (1, 1),
            (1, 1),
        ));
        let con3 = status.register(Conv2d::new(
            40,
            20,
            (2, 2),
            (0, 0),
            nn::Zero,
            (1, 1),
            (1, 1),
        ));
        let con4 = status.register(Conv2d::new(20, 1, (1, 1), (0, 0), nn::Zero, (1, 1), (1, 1)));

        Self {
            con1: con1,
            con2: con2,
            con3: con3,
            con4: con4,
            status,
        }
    }

    fn forward<I, T, U>(
        &self,
        input: I,
    ) -> VarDiff<impl Data<Dim = Ix4>, impl Gradient<Dim = Ix4> + Overwrite>
    where
        I: Convolve<I, Learnable<Ix4>, nn::Zero>,
        I::Output: Into<VarDiff<T, U>>,
        T: Data<Dim = Ix4>,
        U: Gradient<Dim = Ix4> + Overwrite,
    {
        let out1 = self.con1.forward(input).leaky_relu();
        let out2 = self.con2.forward(out1).leaky_relu();
        let out3 = self.con3.forward(out2).leaky_relu();
        let out4 = self.con4.forward(out3);
        out4
    }

    pub fn parameters(&self) -> Vec<Param> {
        self.status.parameters()
    }

    /// Switches the network in training mode.
    fn train(&self) {
        self.status.train()
    }
}

fn main() {
    let model = NeuralNetwork::new();
    let model_optim =
        neuronika::optim::SGD::new(model.parameters(), 0.001, neuronika::optim::L2::new(0.));
    model.train();

    // input data should be irrevelant here (see target below)
    let data_point: Array<f32, Dim<[usize; 4]>> = stack(
        Axis(0),
        &[
            Array::eye(10).view(),
            Array::eye(10).view(),
            Array::eye(10).view(),
            Array::eye(10).view(),
            Array::eye(10).view(),
        ],
    )
    .unwrap()
    .into_shape((1, 5, 10, 10))
    .unwrap();

    for i in 0..600 {
        let x = model.forward(from_ndarray(data_point.clone()));

        // model should learn constant function f(x) = 10
        let y = from_ndarray(Array::from_vec(vec![10.]).into_shape((1, 1, 1, 1)).unwrap());

        let loss = nn::loss::mae_loss(x.clone(), y, nn::loss::Reduction::Mean);

        loss.forward();
        loss.backward(1.);

        println!("epoch {} loss: {}", i, &loss.data());

        model_optim.step();
        model_optim.zero_grad(); // <--- Zeroes the gradients
    }
}
```

Gives:

```
epoch 0 loss: 9.976911
epoch 1 loss: 9.975582
epoch 2 loss: 9.974256
...
epoch 597 loss: 9.320371
epoch 598 loss: 9.319345
epoch 599 loss: 9.3183155
```

**Comment by frjnn** [comment]
Also, you can avoid to call `.train()` on the network since you do not have any dropout layers.

**Comment by lodykas** [comment]
I assumed backward overwrote gradients, but i admit I didn't check. Thanks! 

I see why that would still work with Relu, but why doesn't softplus explode as well, tho?
Well I guess this is not a Neuronika problem, so I'll search on my own about that

**Comment by frjnn** [comment]
`.backward()` cannot overwrite gradients since we also support batch training, gradients are instead accumulated. What were you getting with softplus btw?
